src/hardware/disney.cpp

```diff
@@ -19,456 +20,214 @@
+	channel = MIXER_AddChannel(audio_callback,
+	                           dac_rate_hz,
+	                           "DISNEY",
+	                           {ChannelFeature::Sleep,
+	                            ChannelFeature::ReverbSend,
+	                            ChannelFeature::ChorusSend,
+	                            ChannelFeature::DigitalAudio});
+	assert(channel);
 
-	// Stop and remove the mixer callback
-	if (disney.chan) {
-		disney.chan->Enable(false);
-		disney.chan.reset();
+	if (filter_pref == "on") {
+		// Disney only supports a single fixed 7kHz sample rate. We'll
+		// apply a gentle 6dB/oct LPF at a bit below half the sample
+		// rate to tame the harshest aliased frequencies while still
+		// retaining a good dose of the "raw crunchy DAC sound".
+		constexpr auto lowpass_order = 1;
+		constexpr auto lowpass_freq = static_cast<uint16_t>(dac_rate_hz * 0.45);
+		channel->ConfigureLowPassFilter(lowpass_order, lowpass_freq);
+		channel->SetLowPassFilter(FilterState::On);
+	} else {
+		if (filter_pref != "off")
+			LOG_WARNING("DISNEY: Invalid filter setting '%s', using off",
+			            filter_pref.data());
+		channel->SetLowPassFilter(FilterState::Off);
 	}
 
-	DISNEY_disable(0);
+	// Register port handlers for 8-bit IO
+	const auto write_data = std::bind(&Disney::WriteData, this, _1, _2, _3);
+	write_handler[0].Install(parallel_port, write_data, io_width_t::byte);
+
+	const auto write_control = std::bind(&Disney::WriteControl, this, _1, _2, _3);
+	write_handler[1].Install(parallel_port + 2, write_control, io_width_t::byte);
+
+	const auto read_status = std::bind(&Disney::ReadStatus, this, _1, _2);
+	read_handler.Install(parallel_port + 1, read_status, io_width_t::byte, 2);
+
+	status.power = power_on_bits;
+	LOG_MSG("DISNEY: Disney Sound Source running at %dkHz on LPT1 port %03xh", dac_rate_hz / 1000, parallel_port);
 }
 
-void DISNEY_Init(Section* sec) {
-	Section_prop *section = static_cast<Section_prop *>(sec);
-	assert(section);
-	if (!section->Get_bool("disney"))
-		return;
+Disney::~Disney()
+{
+	LOG_MSG("DISNEY: Shutting down on LPT1 port %03xh", parallel_port);
 
-	// Setup the mixer callback
-	disney.chan = MIXER_AddChannel(DISNEY_CallBack,
-	                               0,
-	                               "DISNEY",
-	                               {ChannelFeature::Sleep,
-	                                ChannelFeature::ReverbSend,
-	                                ChannelFeature::ChorusSend,
-	                                ChannelFeature::DigitalAudio});
+	// Stop the game from accessing the IO ports
+	read_handler.Uninstall();
+	write_handler[0].Uninstall();
+	write_handler[1].Uninstall();
 
-	// Setup zero-order-hold resampler to emulate the "crunchiness" of early
-	// DACs
-	const auto sample_rate = disney.chan->GetSampleRate();
+	channel->Enable(false);
 
-	disney.chan->ConfigureZeroOrderHoldUpsampler(sample_rate);
-	disney.chan->EnableZeroOrderHoldUpsampler();
+	fifo         = {};
```

Comment:
The zero-order-hold resampling stuff definitely should be kept. In combination with the fixed LPF at a bit less than the half of the 7kHz sample rate, it ensures we'll get the raw crunchy sound characteristic of these early DACs, similarly to pre-SB16 SoundBlasters. 

I did post audio samples and measurements of the difference between the mathematically-correct resampling vs the zero-order-hold approach for the DSS, please check it out: https://github.com/dosbox-staging/dosbox-staging/issues/1706#issuecomment-1146731573

So just to be clear, the channel sample rate should be set to the emulated sample rate (7 kHz in this case, and you're already doing that, good!), then the mixer sample rate (e.g. 48 kHz) should be passed to `ConfigureZeroOrderHoldUpsampler` for things to work as intended.

(Now that I think about it, passing the mixer sample rate to `ConfigureZeroOrderHoldUpsampler` is a common scenario, so it could be modified to accept `0` to default to the mixer sample rate.)

Reply:
Thanks for the help! Will give this a go :building_construction: